[WIP] Fix DispatcherTest to test Dispatcher.launchAndDispatch() and Dispatcher's concurrency

kafumi opened the pull request:
## Issue
- close #365

## Overview (Required)
- `Dispatcher` could lose actions sent by `Dispatcher.launchAndDispatch()` in certain cases. The issue was resolved in #364
- Added test functions to test `Dispatcher.launchAndDispatch()` (3c586423150e5ffbcbd3b00800d44c327e9dca4e)
- Also, added other test functions to check all actions are surely dispatched to a subscriber even if the actions are sent concurrently (76bcfb9754deaa669ba33895fc239266301e9ae8)
  - The original version of added tests could FAIL. I applied a temporary modification to pass them (2bfb64ee31702640414ec690b6727fe57b87fc16). It's not good for the test objective, though

## Remaining problem
- Actually it's still possible for `Dispatcher` to lose actions even after #364 is merged
- `Dispatcher` internally uses [`ConflatedBroadcastChannel`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.channels/-conflated-broadcast-channel/) to dispatch actions (elements) to multiple subscribers
  - `ConflatedBroadcastChannel` uses a [`ConflatedChannel`](https://github.com/Kotlin/kotlinx.coroutines/blob/1.0.1/common/kotlinx-coroutines-core-common/src/channels/ConflatedChannel.kt) to send elements to each subscriber
  - `ConflatedChannel` buffers at most one element but its sender never suspends
  - It means, if sending speed exceeds consuming speed, overflowed elements are not buffered and simply lost
    - Only the most recently sent element is kept and passed to subscriber
- I needed to slow down dispatch speed with `delay()` in test functions to make the tests successful (2bfb64ee31702640414ec690b6727fe57b87fc16)
- I tried to replace `ConflatedBroadcastChannel` with `ArrayBroadcastChannel` in `Dispatcher`. But it brought some bugs in app's behavior
  - For example, session data are not loaded when launching app again
  - Following simple fix was not enough for this app...

```diff
diff --git a/frontendcomponent/androidcomponent/src/main/java/io/github/droidkaigi/confsched2019/dispatcher/Dispatcher.kt b/frontendcomponent/androidcomponent/src/main/java/io/github/droidkaigi/confsched2019/dispatcher/Dispatcher.kt
index d99608dd..0dd95c45 100644
--- a/frontendcomponent/androidcomponent/src/main/java/io/github/droidkaigi/confsched2019/dispatcher/Dispatcher.kt
+++ b/frontendcomponent/androidcomponent/src/main/java/io/github/droidkaigi/confsched2019/dispatcher/Dispatcher.kt
@@ -15,7 +15,7 @@ import javax.inject.Singleton

 @Singleton
 class Dispatcher @Inject constructor() {
-    private val _actions = BroadcastChannel<Action>(Channel.CONFLATED)
+    private val _actions = BroadcastChannel<Action>(capacity = 8)
     val events: ReceiveChannel<Action> get() = _actions.openSubscription()

     inline fun <reified T : Action> subscribe(): ReceiveChannel<T> {
```

- In practice, the remaining possibility of action loss doesn't cause obvious bugs for this app
  - I guess it might be rare that multiple actions are sent in very short time 
  - Also, #364 makes sure that `BroadcastChannel.send()` is called in main thread. In actual running circumstance, main thread executes a lot of other stuff including UI events. Those other events possibly prevents `Dispatcher` from occupying main thread to dispatch multiple actions in very short time
    - I mean, the busyness of main thread possibly has effect to slow down the dispatching speed

## Links
N/A

## Screenshot
N/A

kafumi commented:
I'm worried about how we can close #365. I need advice...

takahirom commented:
I want to fix this 😭 But I don't have an idea.

> The original version of added tests could FAIL. I applied a temporary modification to pass them (ba0c2ee). It's not good for the test objective, though

kafumi commented:
Me too... Let me investigate it again after I complete other issues 👓

jmatsu commented:
I guess `Dispatcher#subscribe` is the cause to lost events.

`ReceiveChannel#filter` and `ReceiveChannel#map` will use `RendezvousChannel` as a reciever but it does not have any buffer. It means the channel would drop events in some cases.

```
@UseExperimental(InternalCoroutinesApi::class)
@Singleton
class Dispatcher @Inject constructor() {
    private val _actions = BroadcastChannel<Action>(Channel.CONFLATED)
    val events: ReceiveChannel<Action> get() = _actions.openSubscription()

    inline fun <reified T : Action> subscribe(): ReceiveChannel<T> {
        return events.cast()
    }

    suspend fun dispatch(action: Action) {
        // Make sure calling `_actions.send()` from single thread. We can lose action if
        // `_actions.send()` is called simultaneously from multiple threads
        // https://github.com/Kotlin/kotlinx.coroutines/blob/1.0.1/common/kotlinx-coroutines-core-common/src/channels/ConflatedBroadcastChannel.kt#L227-L230
        withContext(CoroutinePlugin.mainDispatcher) {
            _actions.send(action)
        }
    }

    fun launchAndDispatch(action: Action) {
        GlobalScope.launch(CoroutinePlugin.mainDispatcher) {
            _actions.send(action)
        }
    }

    fun <E, R> ReceiveChannel<E>.cast(
        context: CoroutineContext = Dispatchers.Unconfined
    ): ReceiveChannel<R> =
        GlobalScope.produce(context, capacity = Channel.UNLIMITED, onCompletion = consumes()) {
            consumeEach {
                @Suppress("UNCHECKED_CAST") val e = it as R
                send(e)
            }
        }
}

```

This works fine even if we remove `delay` and increase *numOfActions* to 64 or larger. (I tried to use 256 and it worked for 10 times for now.)

jmatsu commented:
Anyway, we should be careful when using ReceiveChannel's abstractions as much as possible because they are already marked as Obsolete. In addition, all the current abstractions may drop events potentially. 🤔

kafumi commented:
@jmatsu Thank you! I didn't notice the solution that changes the channel type during channel pipeline. I've learned a lot from your suggestion.

I consider `Dispatcher` handles all types of actions on single `BroadcastChannel` and we can't perform unchecked cast on related channels as `ReceiveChannel.cast()` do. I mean, we seem to need `filter()` anyway. But I believe we can apply same approach with using `filter()`. Let me try it and check app's behavior.

Also, regarding `ReceiveChannel`'s obsoleted abstraction, I've never noticed it... 

https://github.com/Kotlin/kotlinx.coroutines/blob/1.1.0/common/kotlinx-coroutines-core-common/src/channels/Channels.common.kt#L632-L633
```kotlin
 * **Note: This API will become obsolete in future updates with introduction of lazy asynchronous streams.**
 *           See [issue #254](https://github.com/Kotlin/kotlinx.coroutines/issues/254).
```

Although I couldn't track all conversation in https://github.com/Kotlin/kotlinx.coroutines/issues/254 (because it's too long...), current channel abstraction doesn't look obsoleted yet. They are thinking of introducing new abstraction method for channels to represent *cold* streams. Current abstractions will become obsolete when the new method is integrated but it's not done yet, I think.
https://github.com/Kotlin/kotlinx.coroutines/issues/254

Anyway, I agree we should be more careful when using channels.

> In addition, all the current abstractions may drop events potentially.

To be honest, I'm not sure about this point. Even though current abstractions are all for *hot* streams, I thought element is not lost unless channel is `ConflatedChannel` because normal channels suspend sender while buffer is full. But I'm not convinced of this my thought now...

Oku ga fukai desu...

kafumi commented on frontendcomponent/androidcomponent/src/test/java/io/github/droidkaigi/confsched2019/dispatcher/DispatcherTest.kt:
```diff
@@ -4,13 +4,17 @@ import io.github.droidkaigi.confsched2019.action.Action
 import kotlinx.coroutines.channels.take
+import kotlinx.coroutines.channels.takeWhile
 import kotlinx.coroutines.channels.toList
+import kotlinx.coroutines.delay
```
Sorry. Amended at 9bebb3f267b06c29276373f60aafe50206d6db30 and force-pushed.

kafumi commented:
Integrated https://github.com/DroidKaigi/conference-app-2019/pull/543#issuecomment-455877736 idea and removed  `delay()` from `DispatcherTest` at 9bebb3f267b06c29276373f60aafe50206d6db30. It works well 🎉 

Also, I checked app's behavior with this fix and latest `master` merged. As far as I roughly checked, there is no problem. I've removed `[WIP]` mark.

jmatsu commented:
> we seem to need filter() anyway.

Oops. Sorry, you are right! I ran only one of testcases so my sample code had no problem, fortunately. I haven't noticed that, thank you!

> To be honest, I'm not sure about this point. Even though current abstractions are all for hot streams, I thought element is not lost unless channel is ConflatedChannel because normal channels suspend sender while buffer is full. But I'm not convinced of this my thought now...

I thought so too but the behavior said *No* to this point. 🤔(BTW *hot* stream isn't related to losing events, I'm thinking.)

> *The comment of ConflatedChannel*
> 
> Channel that buffers at most one element and conflates all subsequent `send` and `offer` invocations,
> so that the receiver always gets the most recently sent element.
> while previously sent elements **are lost**.
> Sender to this channel never suspends and [offer] always returns `true`.

I haven't investigated carefully, so this is just my guess though: the current Channel abstractions say thay do not assure *commutative* and *pure* for now, unfortunately. 😞 

I've tried the code below but events lost ~successfully~ . 

```
@UseExperimental(InternalCoroutinesApi::class)
@Singleton
class Dispatcher @Inject constructor() {
    private val _actions = BroadcastChannel<Action>(Channel.CONFLATED)
    val events: ReceiveChannel<Action> get() = _actions.openSubscription()

    inline fun <reified T : Action> subscribe(): ReceiveChannel<T> {
        return events.castMap()
    }

    inline fun <E, reified R> ReceiveChannel<E>.castMap(
        context: CoroutineContext = Dispatchers.Unconfined
    ): ReceiveChannel<R> =
        GlobalScope.produce(context, capacity = Channel.CONFLATED, onCompletion = consumes()) {
            consumeEach {
                if (it is R) {
                    send(it as R)
                }
            }
        }
}
```

In this case, the source channel of *castMap* is conflated and the destination is also conflated. But events lost anyway. 🤔 

---

> Also, I checked app's behavior with this fix and latest master merged. As far as I roughly checked, there is no problem. I've removed [WIP] mark.

Great! Thank you so much! 🎉

takahirom commented:
I have a question 🙋‍♂️ 
We are using `map` in Store.
Is it lost event?

```
    val loadingState = dispatcher
        .subscribe<Action.SessionLoadingStateChanged>()
        .map { it.loadingState } // ← 
        .toLiveData(LoadingState.INITIALIZED)
```

takahirom commented on frontendcomponent/androidcomponent/build.gradle:
```diff
@@ -43,6 +43,7 @@ repositories {
 tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
     kotlinOptions {
         freeCompilerArgs = [
+                '-Xuse-experimental=kotlin.Experimental',
```
Sorry, I don't understand, What is enabled by `kotlin.Experimental`?

jmatsu replied:
this one. https://github.com/DroidKaigi/conference-app-2019/pull/543/files#diff-8b8eae244edcf205243c530896a91894R44

kafumi replied:
I just noticed we don't need to use this internal version of `CoroutineScope.produce()`. I mean,  we don't need `@UseExperimental(InternalCoroutinesApi::class)` annotation and `-Xuse-experimental=kotlin.Experimental` flag actually. Let me fix them :bow:

There are two versions of `produce()` and the internal version have an additional parameter `onCompletion`. `onCompletion` is used as the `CompletionHandler` of the `produce`'s coroutine.

- Public version: https://github.com/Kotlin/kotlinx.coroutines/blob/1.1.0/common/kotlinx-coroutines-core-common/src/channels/Produce.kt#L60-L64
- Internal version: https://github.com/Kotlin/kotlinx.coroutines/blob/1.1.0/common/kotlinx-coroutines-core-common/src/channels/Produce.kt#L76-L82

`map()` and `filter()` uses the internal version of `produce()` to set [`ReceiveChannel.consumes()`](https://github.com/Kotlin/kotlinx.coroutines/blob/1.1.0/common/kotlinx-coroutines-core-common/src/channels/Channels.common.kt#L62-L66) to `onCompletion`. It's needed for the case in which `CoroutineContext`'s job is cancelled before `map()`/`filter()` consumes all elements. It ensures all elements in `ReceiveChannel` are consumed.

In our case, we use `GlobalScope` without setting custom `CoroutineContext`'s elements to call `produce()` in `Dispatcher`. It means the `CompletionHandler` should be never called. We can safely use the public version of `produce()` which don't have `onCompletion` parameter.

kafumi replied:
Fixed at 94e539b45afd4f82b81d60450e0bbd95628ca359.

kafumi commented:
@jmatsu I totally agree that it's possible to lose elements with `ConflatedChannel` because it never suspend sender and receiver as you referred. If sender's speed exceeds receiver's speed, elements are overflowed and lost. Sorry, I tried to write so in my previous comment, but it was too short and not clear...
> I thought element is not lost **unless** channel is ConflatedChannel

---

@takahirom Regarding your last question, I believe (as far as I understand) any operations at `Dispatcher`'s subscribers are now safe because we have an unlimited buffer between `BroadcastChannel` and `Dispatcher`'s interface. If subscribers' consume speed is low, elements are buffered at the intermediate channel created by `GlobalScope.produce()` below.

```kotlin
    inline fun <reified E, reified R : E> ReceiveChannel<E>.filterAndCast(
        context: CoroutineContext = Dispatchers.Unconfined
    ): ReceiveChannel<R> =
        GlobalScope.produce(context, capacity = Channel.UNLIMITED, onCompletion = consumes()) {
```
